dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,204 @@
+
+instance Monoid Tags where
+    mempty = Tags M.empty
+
+data Tag = Tag
+    { tagDefinition :: Text
+      -- ^ In vtags source code this field has name pattern and EMacs use it as regex pattern
+      --   to locate line with tag. It's looking for ^<tag definition>.
+      --   Looks like vi is not using it.
+    , tagName :: Text
+      -- ^ text, that editor compare with selected text. So it's really name of entity
+    }
+
+{-| Generate ETags for Dhall expressions
+-}
+generate
+    :: Input
+    -- ^ Path to directory or filename that should be indexed
+    -> [Text]
+    -- ^ List of suffixes for dhall files [".dhall"] by default
+    -> Bool
+    -- ^ Flag if generate should follow symlinks
+    -> IO Text
+    -- ^ Content for tags file
+generate inp sxs followSyms = do
+    isD <- case inp of
+               StandardInput -> return False
+               InputFile p -> doesDirectoryExist p
+    files <- inputToFiles followSyms (map T.unpack (if isD then sxs else [""])) inp
+    tags <- foldMap (\f -> handle (\(SomeException _) -> return mempty)
+                                  (fileTags f <$> TIO.readFile f)) files
+    return (showTags tags)
+
+{-| Find tags in Text (second argument) and generates list of them
+    To make tags for filenames working in both, emacs and vi, add two initial tags.
+    First for `filename` for vi and second with `/filename` for emacs.
+    Other tags are working for both.
+-}
+fileTags :: FilePath -> Text -> Tags
+fileTags f t = Tags (M.singleton f
+                    (initialMap <> getTagsFromText t))
+    where initialViTag = (FP 1 0, Tag "" (T.pack . takeFileName $ f))
+          initialEmacsTag = (FP 1 1, Tag "" ("/" <> (T.pack . takeFileName) f))
+          initialMap = M.fromList [initialViTag, initialEmacsTag]
+
+getTagsFromText :: Text -> M.Map PosInFile Tag
+getTagsFromText t = fromRight M.empty $
+                              fixPosAndDefinition t . getTagsFromExpr <$> exprFromText "" t
+
+-- after getTagsFromExpr line and column in line are in PosInFile for each tag
+-- and tag definition is empty.
+-- Emacs use tag definition to check if tag is on line. It compares line from start
+-- with tag definition and in case they are the same, relocate user.
+-- fixPosAndDefinition change position to line and byte offset and add tag definition.
+fixPosAndDefinition :: Text -> M.Map PosInFile Tag -> M.Map PosInFile Tag
+fixPosAndDefinition t = M.foldMapWithKey (\(FP ln c) (Tag _ term) ->
+             let (lsl, l) = fromMaybe (0, "") (ln `M.lookup` mls) in
+             M.singleton (FP ln (lsl + c)) (Tag (T.take c l) term))
+    where mls :: M.Map Int (Int, Text) 
+          -- mls is map that for each line has length of file before this map and line content.
+          mls = M.fromList . fst .
+                foldl (\(ls, lsl) (n, l) ->
+                           let lsl' = lsl + 1 + T.length l in
+                           ((n, (lsl, l)):ls, lsl')) ([], 0) . zip [1..] $ T.lines t
```

Comment:
I make a lot of fixes her, but actually, algorithm is even more complex now.
So sorry, it's possible that it's even more difficult to understand now.
I wrote comments here. I hope, it allow anyone to understand it.
But in case it's still to messy, I'll think how to change it. If you have any suggestion here, I'll be happy to implement them.

Reply:
Even if I'll do it the same way as it's in hasktags, I still need to find proper line. So it will make logic simpler, but still not very simple.
I can replace Map with Vector. It will make it faster, but not simpler.
Are you sure, if I need to remove byte offset calculation here?

Reply:
I think at this point it would be easier to accept the change and follow-up with another pull request of your own to refactor the code, especially given that this has test coverage.  It's easier to communicate what style you prefer through a pull request of your own (which @DKurilo can then review!) than to try to communicate the preferred style via issue comments

Reply:
@sjakobi , I tried to make it more clear.